kolibri/plugins/learn/assets/src/views/CustomContentRenderer.vue

```diff
@@ -0,0 +1,223 @@
+        this.fetchContentModel(message);
+      });
+      this.hashi.on('navigateto', message => {
+        this.navigateTo(message);
+      });
+      this.hashi.on('context', message => {
+        this.getOrUpdateContext(message);
+      });
+    },
+    methods: {
+      // helper functions for fetching data from kolibri
+      // called in mainClient.js
+
+      fetchContentCollection(message) {
+        const options = message.options;
+        const getParams = pick(options, ['ids', 'parent']);
+        if (options.parent && options.parent == 'self') {
+          getParams.parent = this.topic.id;
+        }
+        let self = this;
+        ContentNodeResource.fetchCollection({ getParams }).then(contentNodes => {
+          contentNodes ? (message.status = 'success') : (message.status = 'failure');
+          let response = {};
+          response.page = message.options.page ? message.options.page : 1;
+          response.pageSize = message.options.pageSize ? message.options.pageSize : 50;
+          response.results = contentNodes;
+          message.data = response;
+          message.type = 'response';
+          self.hashi.mediator.sendLocalMessage({
+            nameSpace: 'hashi',
+            event: events.KOLIBRIDATARETURNED,
+            data: message,
+          });
+        });
+      },
+
+      fetchContentModel(message) {
+        let id = message.id;
+        let self = this;
+        ContentNodeResource.fetchModel({ id }).then(contentNode => {
+          if (contentNode) {
+            message.status = 'success';
+          } else {
+            message.status = 'failure';
+          }
+          message.data = contentNode;
+          message.type = 'response';
+          self.hashi.mediator.sendMessage({
+            nameSpace: 'hashi',
+            event: events.KOLIBRIDATARETURNED,
+            data: message,
+          });
+        });
+      },
+
+      navigateTo(message) {
+        let id = message.nodeId;
+        let self = this;
+        ContentNodeResource.fetchModel({ id }).then(contentNode => {
+          let routeBase, path;
+          if (contentNode && contentNode.kind === 'topic') {
+            routeBase = '/topics/t';
+            path = `${routeBase}/${id}`;
+            router.push({ path: path }).catch(() => {});
+          } else if (contentNode) {
+            // in a custom context, launch or maintain overlay
+            self.currentContent = contentNode;
+            // self.open = true;
+          }
+          self.hashi.mediator.sendLocalMessage({
+            nameSpace: 'hashi',
+            event: events.KOLIBRIDATARETURNED,
+            data: message,
+          });
+        });
+      },
+      getOrUpdateContext(message) {
+        // to update context with the incoming context
+        if (message.context) {
+          message.context.customChannel = message.context.customChannel || true;
+          const encodedContext = JSON.stringify(this.encodeContext(message.context));
```

Comment:
I think this is the wrong way around - would want to stringify the context object first, then URI encode it (can also just do both inside the `encodeContext` method.